docs/the-new-architecture/pillars-cxxturbomodules.md

```diff
@@ -0,0 +1,698 @@
+
+## Supporting custom structs
+
+You can use the same approach for you custom types in JavaScript such as this one:
+
+```js
+export type CustomType = {
+  key: string,
+  enabled: boolean,
+  time?: number
+};
+```
+
+which can be exposed to your C++ Turbo Native Module via
+
+<Tabs groupId="turbomodule-specs" defaultValue={constants.defaultJavaScriptSpecLanguages} values={constants.javaScriptSpecLanguages}>
+<TabItem value="typescript">
+
+```typescript title="NativeSampleModule.ts"
+   // ...
+   readonly passCustomType: (input: CustomType) => CustomType;
+   // ..
+```
+
+</TabItem>
+<TabItem value="flow">
+
+```js title="NativeSampleModule.js"
+   // ...
+   +passCustomType: (input: CustomType) => CustomType;
+   // ..
+```
+
+</TabItem>
+</Tabs>
+
+To use this custom type in C++, you need to define your custom Struct and bridging function e.g., directly in `NativeSampleModule.h`:
+
+```cpp
+struct CustomType {
+  std::string key;
+  bool enabled;
+  std::optional<int32_t> time;
+
+  bool operator==(const CustomType &other) const {
+    return key == other.key && enabled == other.enabled && time == other.time;
+  }
+};
+
+template <>
+struct Bridging<CustomType> {
+  static CustomType fromJs(
+      jsi::Runtime &rt,
+      const jsi::Object &value,
+      const std::shared_ptr<CallInvoker> &jsInvoker) {
+    return CustomType{
+        .key = bridging::fromJs<std::string>(
+            rt, value.getProperty(rt, "key"), jsInvoker),
+        .enabled = bridging::fromJs<bool>(
+            rt, value.getProperty(rt, "enabled"), jsInvoker),
+        .time = bridging::fromJs<std::optional<int32_t>>(
+            rt, value.getProperty(rt, "time"), jsInvoker)};
+  }
+
+  static jsi::Object toJs(jsi::Runtime &rt, const CustomType &value) {
+    auto result = facebook::jsi::Object(rt);
+    result.setProperty(rt, "key", bridging::toJs(rt, value.key));
+    result.setProperty(rt, "enabled", bridging::toJs(rt, value.enabled));
+    if (value.time) {
+      result.setProperty(rt, "time", bridging::toJs(rt, value.time.value()));
+    }
+    return result;
+  }
+};
+```
+
+Declare it in your `NativeSampleModule.h` file:
+
+```cpp
+  CustomType passCustomType(jsi::Runtime& rt, CustomType input);
+```
+
+Implement it in `NativeSampleModule.cpp`:
+
+```cpp
+CustomType NativeSampleModule::passCustomType(jsi::Runtime& rt, CustomType input) {
+  input.key = "1909";
+  input.enabled = !input.enabled;
+  input.time = 42;
+  return input;
+}
+```
+
+In your app you can call this new native function via:
+
+```js
+<Section title="Cxx TurboModule">
+  NativeSampleModule.passCustomType(...) ={' '}
+  {JSON.stringify(
+    NativeSampleModule.passCustomType({
+      key: '123',
+      enabled: true,
+      time: undefined
+    })
```

Comment:
Don't we lose type safety here?

Reply:
We don't.

Here is the spec
`
 export type CustomType = {
   key: string;
   enabled: boolean;
   time?: number;
 }
`

Flow/TypeScript checker will complain if you pass in valid data here